[render-blocking] Render blocking for <link> should not match elements that are on a 'stack of open elements' for the parser.

mattwoodrow opened the pull request:
#### fb62715ca67c53d52c82cca2eafe460777a2a4ae
<pre>
[render-blocking] Render blocking for &lt;link&gt; should not match elements that are on a &apos;stack of open elements&apos; for the parser.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279167">https://bugs.webkit.org/show_bug.cgi?id=279167</a>
&lt;<a href="https://rdar.apple.com/135846827">rdar://135846827</a>&gt;

Reviewed by Tim Nguyen.

Check if the Element being considered for matching a &lt;link rel=expect&gt; is still
on the stack of open elements from the parser.

This requires adding another call to processInternalResourceLink when an Element
is popped from the stack, and making sure that finishParsingChildren is called
*after* the stack gets mutated (otherwise recursive calls to check the stack
think its still there).

Generalizes the fast path of processInternalResourceLink (where a specific
Element to check is provided) to accept any Element instead of only anchors, and
uses it for the stack pop case.

Pass the Element being mutated into IdTargetObserver, so that this path too can
use the provided Element fast path check. We should now only do the tree-walk
search for an Element when the &lt;link&gt; itself is mutated.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/element-render-blocking-040-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/element-render-blocking-040.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::htmlDocumentParser const):
(WebCore::Document::processInternalResourceLinks):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentParser.h:
(WebCore::DocumentParser::asHTMLDocumentParser):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):
(WebCore::Element::finishParsingChildren):
* Source/WebCore/dom/IdTargetObserver.h:
* Source/WebCore/dom/IdTargetObserverRegistry.cpp:
(WebCore::IdTargetObserverRegistry::notifyObserversInternal):
* Source/WebCore/dom/IdTargetObserverRegistry.h:
(WebCore::IdTargetObserverRegistry::notifyObservers):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::addElementById):
(WebCore::TreeScope::removeElementById):
(WebCore::TreeScope::findAnchor):
(WebCore::TreeScope::isMatchingAnchor):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::FormAttributeTargetObserver::idTargetChanged):
* Source/WebCore/html/HTMLDataListElement.cpp:
(WebCore::HTMLDataListElement::optionElementChildrenChanged):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::ListAttributeTargetObserver::idTargetChanged):
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::ExpectIdTargetObserver::idTargetChanged):
(WebCore::HTMLLinkElement::processInternalResourceLink):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::isOnStackOfOpenElements const):
* Source/WebCore/html/parser/HTMLDocumentParser.h:
* Source/WebCore/html/parser/HTMLElementStack.cpp:
(WebCore::HTMLElementStack::popAll):
(WebCore::HTMLElementStack::popCommon):
(WebCore::HTMLElementStack::removeNonTopCommon):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::isOnStackOfOpenElements const):
* Source/WebCore/html/parser/HTMLTreeBuilder.h:

Canonical link: <a href="https://commits.webkit.org/289020@main">https://commits.webkit.org/289020@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d605a1a8704b1178b089174eb3380c13367bb29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84974 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4703 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39370 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90123 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36030 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12679 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66128 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23944 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88019 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3665 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77224 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46395 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3544 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31451 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35103 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74349 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32261 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91496 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12318 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9002 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74612 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73034 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73732 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18266 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18121 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16573 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4342 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12266 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17719 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12102 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15598 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13847 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/element-render-blocking-040.html:
There's actually no -039.html upstream!

mattwoodrow replied:
I know :(

I tried to export it, but it fails lint, which is unfortunate.

I'll make a new PR to fixup -039, and then do a manual export of the final version.

mattwoodrow replied:
Done!

-40 is now the next free number :)

mattwoodrow committed:
[render-blocking] Render blocking for <link> should not match elements that are on a 'stack of open elements' for the parser.
https://bugs.webkit.org/show_bug.cgi?id=279167
<rdar://135846827>

Reviewed by Tim Nguyen.

Check if the Element being considered for matching a <link rel=expect> is still
on the stack of open elements from the parser.

This requires adding another call to processInternalResourceLink when an Element
is popped from the stack, and making sure that finishParsingChildren is called
*after* the stack gets mutated (otherwise recursive calls to check the stack
think its still there).

Generalizes the fast path of processInternalResourceLink (where a specific
Element to check is provided) to accept any Element instead of only anchors, and
uses it for the stack pop case.

Pass the Element being mutated into IdTargetObserver, so that this path too can
use the provided Element fast path check. We should now only do the tree-walk
search for an Element when the <link> itself is mutated.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/element-render-blocking-040-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/element-render-blocking-040.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::htmlDocumentParser const):
(WebCore::Document::processInternalResourceLinks):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentParser.h:
(WebCore::DocumentParser::asHTMLDocumentParser):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):
(WebCore::Element::finishParsingChildren):
* Source/WebCore/dom/IdTargetObserver.h:
* Source/WebCore/dom/IdTargetObserverRegistry.cpp:
(WebCore::IdTargetObserverRegistry::notifyObserversInternal):
* Source/WebCore/dom/IdTargetObserverRegistry.h:
(WebCore::IdTargetObserverRegistry::notifyObservers):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::addElementById):
(WebCore::TreeScope::removeElementById):
(WebCore::TreeScope::findAnchor):
(WebCore::TreeScope::isMatchingAnchor):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::FormAttributeTargetObserver::idTargetChanged):
* Source/WebCore/html/HTMLDataListElement.cpp:
(WebCore::HTMLDataListElement::optionElementChildrenChanged):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::ListAttributeTargetObserver::idTargetChanged):
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::ExpectIdTargetObserver::idTargetChanged):
(WebCore::HTMLLinkElement::processInternalResourceLink):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::isOnStackOfOpenElements const):
* Source/WebCore/html/parser/HTMLDocumentParser.h:
* Source/WebCore/html/parser/HTMLElementStack.cpp:
(WebCore::HTMLElementStack::popAll):
(WebCore::HTMLElementStack::popCommon):
(WebCore::HTMLElementStack::removeNonTopCommon):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::isOnStackOfOpenElements const):
* Source/WebCore/html/parser/HTMLTreeBuilder.h:

Canonical link: https://commits.webkit.org/289020@main
